File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def __init__(self):
        self.ans = []

    def solve(self, idx, curr, arr, sum):
        if sum == 0:
            self.ans.append(list(curr))  
            return

        for i in range(idx, len(arr)):
            if i > idx and arr[i] == arr[i - 1]:  # Skip duplicates
                continue
            if sum < arr[i]:  
                break

            curr.append(arr[i])
            self.solve(i + 1, curr, arr, sum - arr[i])  
            curr.pop()  

    def combinationSum2(self, arr, target):
        arr.sort()  # Sort the array to handle duplicates
        self.solve(0, [], arr, target)
        return self.ans

        

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> ans;
    void solve(int idx, vector<int> &curr, vector<int> &arr, int sum) {
        int n = arr.size();
        if(sum == 0) {
            ans.push_back(curr);
            return;
        }

        for(int i = idx; i<n; i++) {
            if(i > idx && arr[i] == arr[i-1]) continue;
            if(sum < arr[i]) break;
            curr.push_back(arr[i]);
            solve(i + 1, curr, arr, sum - arr[i]);
            curr.pop_back();
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& arr, int target) {
        vector<int> curr;
        sort(arr.begin(), arr.end());
        solve(0, curr, arr, target);
        return ans;
    }
};